File: python/735.py
"""Project Euler Problem 735: Divisors of 2n².

Let f(n) be the number of divisors of 2n² no greater than n. Find Σ_{k=1}^N f(k).

Each divisor can be represented as the first factor of the product
(n-x)(2n+y)=2n², where 0≤x≤n. This can be rearranged to n=xy/(y-2x). Let
g = GCD(x,y), x = g*x', and y = g*y'. Then for any relatively prime (x',y'),
first we must have y'>2x' in order for n to be positive. Further, we must have
y'-2x'|2g, so we need to choose g such that n=g²x'y'/(g(y'-2x')) = k*x'y' ≤ N,
where k is an integer or half-integer if y'-2x' is even. This yields
⌊N/(x'y')⌋ or ⌊2N/(x'y')⌋ solutions for a given (x',y').

We remove the GCD constraint in the usual way. For the half-integer constraint,
we can encode the parity constraint by subtracting solutions with N replaced
with N/2, then adding back it with N/4, etc. The problem then becomes
computing Σ_{(x',y')} ⌊N/(x'y')⌋, for both y'>x' and y'>2x', which is equal to
the number of solutions to x*y*z ≤ N with each of these constraints. To
compute this efficiently in O(N^{2/3}) time, we iterate over the smaller two
of (x,y,z), and count the number of solutions for the third variable.
"""

from __future__ import annotations

from math import gcd, isqrt


def pre_mobius(limit: int) -> list[int]:
    """Precompute Mobius function."""
    mobius = [1] * (limit + 1)
    is_prime = [True] * (limit + 1)
    is_prime[0] = is_prime[1] = False

    for i in range(2, limit + 1):
        if is_prime[i]:
            for j in range(i, limit + 1, i):
                is_prime[j] = False
                mobius[j] *= -1
            for j in range(i * i, limit + 1, i * i):
                mobius[j] = 0

    return mobius


def sq(n: int) -> int:
    """Square."""
    return n * n


def cb(n: int) -> int:
    """Cube."""
    return n * n * n


def parity(n: int) -> int:
    """Return 1 if n is even, -1 if odd."""
    return 1 if n % 2 == 0 else -1


def solve() -> int:
    """Solve Problem 735."""
    N = 10**12
    L = isqrt(N)

    mobius = pre_mobius(L)
    ans = N

    for g in range(1, L + 1):
        if sq(g) >= N:
            break
        t = 0
        while sq(g) << t <= N:
            n_val = N // sq(g) >> t

            res = 0

            # x*y*z≤n, y>x
            for x in range(1, int(n_val ** (1 / 3)) + 1):
                if cb(x) > n_val:
                    break
                for y in range(x + 1, int((n_val / x) ** 0.5) + 1):
                    if x * sq(y) > n_val:
                        break
                    res += n_val // (x * y) - y

            for x in range(1, int(n_val ** (1 / 3)) + 1):
                if cb(x) > n_val:
                    break
                for z in range(x + 1, int((n_val / x) ** 0.5) + 1):
                    if x * sq(z) > n_val:
                        break
                    res += n_val // (x * z) - (z - 1)

            for z in range(1, int(n_val ** (1 / 3)) + 1):
                if cb(z) > n_val:
                    break
                for x in range(z, int((n_val / z) ** 0.5) + 1):
                    if z * sq(x) > n_val:
                        break
                    res += n_val // (x * z) - x

            # x*y*z≤n, y>2x
            for x in range(1, int(n_val ** (1 / 3)) + 1):
                if cb(x) > n_val:
                    break
                for y in range(2 * x + 1, int((n_val / x) ** 0.5) + 1):
                    if x * sq(y) > n_val:
                        break
                    res += n_val // (x * y) - y

            for x in range(1, int(n_val ** (1 / 3)) + 1):
                if cb(x) > n_val:
                    break
                for z in range(x + 1, int((n_val / x) ** 0.5) + 1):
                    if x * sq(z) > n_val or 2 * z * sq(x) > n_val:
                        break
                    res += n_val // (x * z) - max(2 * x, z - 1)

            for z in range(1, int(n_val ** (1 / 3)) + 1):
                if cb(z) > n_val:
                    break
                for x in range(z, int((n_val / (2 * z)) ** 0.5) + 1):
                    if 2 * z * sq(x) > n_val:
                        break
                    res += n_val // (x * z) - 2 * x

            ans += res * parity(t) * mobius[g]
            t += 1

    return ans


def main() -> int:
    """Main entry point."""
    result = solve()
    print(result)
    return result


if __name__ == "__main__":
    main()

File: c/735.c
/* Project Euler 735: Divisors of 2n^2.
 * Mobius function sieve + counting lattice points in O(N^{2/3}) per term.
 */
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <stdint.h>

typedef long long ll;

#define LIMIT 1000001  /* sqrt(10^12) + 1 */

static int mobius[LIMIT];

void pre_mobius(int limit) {
    static char is_prime[LIMIT];
    for (int i = 0; i <= limit; i++) { mobius[i] = 1; is_prime[i] = 1; }
    is_prime[0] = is_prime[1] = 0;

    for (int i = 2; i <= limit; i++) {
        if (is_prime[i]) {
            for (int j = i; j <= limit; j += i) {
                if (j != i) is_prime[j] = 0;
                mobius[j] *= -1;
            }
            for (ll j = (ll)i * i; j <= limit; j += (ll)i * i) {
                mobius[j] = 0;
            }
        }
    }
}

static inline ll isq(ll n) { return n * n; }
static inline ll icb(ll n) { return n * n * n; }

int main() {
    ll N = 1000000000000LL; /* 10^12 */
    ll L = (ll)sqrt((double)N);
    if (L * L > N) L--;
    while ((L+1)*(L+1) <= N) L++;

    pre_mobius((int)L);

    ll ans = N;

    for (ll g = 1; isq(g) < N; g++) {
        if (mobius[g] == 0) continue;

        int t = 0;
        while (isq(g) * (1LL << t) <= N) {
            ll n_val = (N / isq(g)) >> t;
            if (n_val < 1) { t++; continue; }

            ll res = 0;

            /* x*y*z <= n_val, y > x */
            {
                ll cbrt_n = (ll)cbrt((double)n_val);
                while (icb(cbrt_n + 1) <= n_val) cbrt_n++;
                while (icb(cbrt_n) > n_val) cbrt_n--;

                for (ll x = 1; icb(x) <= n_val; x++) {
                    ll nox = n_val / x;
                    ll sq_nox = (ll)sqrt((double)nox);
                    while (sq_nox * sq_nox > nox) sq_nox--;
                    while ((sq_nox+1)*(sq_nox+1) <= nox) sq_nox++;

                    for (ll y = x + 1; y <= sq_nox; y++) {
                        res += n_val / (x * y) - y;
                    }
                }

                for (ll x = 1; icb(x) <= n_val; x++) {
                    ll nox = n_val / x;
                    ll sq_nox = (ll)sqrt((double)nox);
                    while (sq_nox * sq_nox > nox) sq_nox--;
                    while ((sq_nox+1)*(sq_nox+1) <= nox) sq_nox++;

                    for (ll z = x + 1; z <= sq_nox; z++) {
                        res += n_val / (x * z) - (z - 1);
                    }
                }

                for (ll z = 1; icb(z) <= n_val; z++) {
                    ll noz = n_val / z;
                    ll sq_noz = (ll)sqrt((double)noz);
                    while (sq_noz * sq_noz > noz) sq_noz--;
                    while ((sq_noz+1)*(sq_noz+1) <= noz) sq_noz++;

                    for (ll x = z; x <= sq_noz; x++) {
                        res += n_val / (x * z) - x;
                    }
                }
            }

            /* x*y*z <= n_val, y > 2x */
            {
                for (ll x = 1; icb(x) <= n_val; x++) {
                    ll nox = n_val / x;
                    ll sq_nox = (ll)sqrt((double)nox);
                    while (sq_nox * sq_nox > nox) sq_nox--;
                    while ((sq_nox+1)*(sq_nox+1) <= nox) sq_nox++;

                    for (ll y = 2 * x + 1; y <= sq_nox; y++) {
                        res += n_val / (x * y) - y;
                    }
                }

                for (ll x = 1; icb(x) <= n_val; x++) {
                    ll nox = n_val / x;
                    ll sq_nox = (ll)sqrt((double)nox);
                    while (sq_nox * sq_nox > nox) sq_nox--;
                    while ((sq_nox+1)*(sq_nox+1) <= nox) sq_nox++;

                    for (ll z = x + 1; z <= sq_nox; z++) {
                        if (2 * z * isq(x) > n_val) break;
                        ll maxv = 2 * x;
                        if (z - 1 > maxv) maxv = z - 1;
                        res += n_val / (x * z) - maxv;
                    }
                }

                for (ll z = 1; icb(z) <= n_val; z++) {
                    ll half_noz = n_val / (2 * z);
                    ll sq_half = (ll)sqrt((double)half_noz);
                    while (sq_half * sq_half > half_noz) sq_half--;
                    while ((sq_half+1)*(sq_half+1) <= half_noz) sq_half++;

                    for (ll x = z; 2 * z * isq(x) <= n_val; x++) {
                        res += n_val / (x * z) - 2 * x;
                    }
                }
            }

            int parity = (t % 2 == 0) ? 1 : -1;
            ans += res * parity * mobius[g];
            t++;
        }
    }

    printf("%lld\n", ans);
    return 0;
}
